src/qubes_fwupdmgr.py

```diff
@@ -0,0 +1,502 @@
+                            "Checksum": update["Checksum"][0],
+                            "Description": update["Description"]
+                        } for update in device["Releases"]
+                    ]
+            } for device in self.updates_info_dict["Devices"]
+        ]
+
+    def _download_firmware_updates(self, url, sha):
+        """Initializes downloading firmware upadate archive.
+
+        Keywords arguments:
+        url -- url path to firmware upadate archive
+        sha -- SHA1 checksum of the firmware update archive
+        """
+        name = url.replace("https://fwupd.org/downloads/", "")
+        update_path = os.path.join(
+            FWUPD_DOM0_UPDATES_DIR,
+            name.replace(".cab", "")
+        )
+        cmd_fwdownload = [
+            FWUPD_DOM0_UPDATE,
+            "--update",
+            "--url=%s" % url,
+            "--sha=%s" % sha
+        ]
+        p = subprocess.Popen(cmd_fwdownload)
+        p.wait()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Firmware download failed")
+        if not os.path.exists(update_path):
+            raise ValueError("Firmware update files do not exist")
+
+    def _user_input(self, updates_list, downgrade=False):
+        """UI for update process.
+
+        Keywords arguments:
+        updates_list - list of updates for specified device
+        downgrade -- downgrade flag
+        """
+        decorator = "======================================================"
+        if len(updates_list) == 0:
+            print("No updates available.")
+            return 99
+        if downgrade:
+            print("Available downgrades:")
+        else:
+            print("Available updates:")
+        print(decorator)
+        for i, device in enumerate(updates_list):
+            print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
+            print("%s. Device: %s" % (i+1, device["Name"]))
+            print("   Current firmware version:\t %s" % device["Version"])
+            for update in device["Releases"]:
+                print(decorator)
+                print("   Firmware update version:\t %s" % update["Version"])
+                description = update["Description"].replace("<p>", "")
+                description = description.replace("<li>", "")
+                description = description.replace("<ul>", "")
+                description = description.replace("</ul>", "")
+                description = description.replace("</p>", "\n   ")
+                description = description.replace("</li>", "\n   ")
+                print("   Description: %s" % description)
+            print(decorator)
+
+        while True:
+            try:
+                print("If you want to abandon process press 'N'.")
+                choice = input(
+                    "Otherwise choose a device number: "
+                )
+                if choice == 'N' or choice == 'n':
+                    return 99
```

Comment:
Can we define a dict with possible exit codes and name them somehow, rather than hardocding integers

Reply:
Done.